tests/apollo/util/bft.py

```diff
@@ -630,18 +630,9 @@ def force_quorum_including_replica(self, replica_id, primary=0):
         so that the remaining replicas form a quorum that includes replica_id
         """
         with log.start_action(action_type="force_quorum_including_replica") as action:
-            unstable_replicas = self.all_replicas(without={primary, replica_id})
-
-            random.shuffle(unstable_replicas)
-
-            for backup_replica_id in unstable_replicas:
-                action.log(f'Stopping backup replica {backup_replica_id} in order '
-                      f'to force a quorum including replica {replica_id}...')
-                self.stop_replica(backup_replica_id)
-                if len(self.procs) == 2 * self.config.f + self.config.c + 1:
-                    break
-
-            assert len(self.procs) == 2 * self.config.f + self.config.c + 1
+            assert len(self.procs) >= 2 * self.config.f + self.config.c + 1
+            self.stop_replicas(self.random_set_of_replicas(
+                len(self.procs) - (2 * self.config.f + self.config.c + 1), without={primary, replica_id}))
```

Comment:
why not just 
```python
            self.stop_replicas(self.random_set_of_replicas(f, without={primary, replica_id}))
```
?

Reply:
The original code was correct. We need at least 2f + c + 1 nodes for the quorum, so for example in the edge case n6 c1 f1 we need to stop 2 instead of 1 to ensure the relevant replica_id is a non failed part of the quorum of 4.

Reply:
ok, makes sense